### PR TITLE
BOAC-2051: Implements add/remove topic feature for note creation

### PR DESCRIPTION
**Pull request by lyttam**
https://jira.ets.berkeley.edu/jira/browse/BOAC-2051

Also fixes some apparent regressions that should have been included in the previous PR.

### CODE REVIEW

File: boac/api/notes_controller.py [hunk 1]
```diff
@@ -201,7 +201,7 @@ def _get_name(user):
 
 def _get_topics(params):
     topics = params.get('topics', ())
-    return topics if isinstance(topics, list) else str(topics).split(',')
+    return topics if isinstance(topics, list) else list(filter(None, str(topics).split(',')))
```
Comment (johncrossman) [inline boac/api/notes_controller.py]: @lyttam Why doesn't `split()` return a list as you need?

File: tests/test_api/test_notes_controller.py [hunk 2]
```diff
@@ -120,6 +120,9 @@ def test_create_note_with_topics(self, app, client, fake_auth):
             topics=['collaborative synergies', 'integrated architectures', 'vertical solutions'],
         )
         assert len(note.get('topics')) == 3
+        assert note.get('topics')[0] == 'COLLABORATIVE SYNERGIES'
+        assert note.get('topics')[1] == 'INTEGRATED ARCHITECTURES'
+        assert note.get('topics')[2] == 'VERTICAL SOLUTIONS'
```
Comment (johncrossman) [inline tests/test_api/test_notes_controller.py]: @lyttam Will we be displaying topics in all caps on front-end?

File: src/api/notes.ts [hunk 3]
```diff
@@ -10,11 +10,18 @@ export function markRead(noteId) {
...
+    body: string,
+    topics: string[],
+    attachments: any[]
+) {
   const data = {
     sid: sid.toString(),
     subject: subject,
-    body: body
+    body: body,
+    topics: topics,
```
Comment (johncrossman) [inline src/api/notes.ts]: Minor: trailing comma above

File: tests/test_api/test_notes_controller.py (hunk 2 above)
Comment (lyttam) [reply tests/test_api/test_notes_controller.py]: Yes, according to the mockup. But I also made sure to store them in the database in all caps to facilitate future autocomplete functionality. I think I'd rather have them be in title case so I might change that later, but for now all caps was the easiest thing to do.

File: src/components/note/NewNoteModal.vue [hunk 4]
```diff
@@ -79,6 +79,53 @@
...
+                        class="px-0 pt-1"
+                        @click.prevent="removeTopic(topic)">
+                        <i class="fas fa-times-circle has-error pl-2"></i>
+                      </b-btn>
+                    </span>
+                  </li>
+                </ul>
+              </div>
+            </b-container>
+          </div>
```
Comment (johncrossman) [inline src/components/note/NewNoteModal.vue]: @lyttam the `b-container` above appears to be identical to what's in EditAdvisingNote. Would a shared component make sense?

File: boac/api/notes_controller.py (hunk 1 above)
Comment (lyttam) [reply boac/api/notes_controller.py]: It does, but if a blank topic is submitted the list will contain an empty string. The purpose of `filter` is to remove the empty strings. Is there a better way to do this?

File: tests/test_api/test_notes_controller.py (hunk 2 above)
Comment (johncrossman) [reply tests/test_api/test_notes_controller.py]: Thanks @lyttam  The mockups in BOAC-1924 have title case. Has there been a revision of the design? IMO, all caps in the UI would be a bit jarring.

File: boac/api/notes_controller.py (hunk 1 above)
Comment (johncrossman) [reply boac/api/notes_controller.py]: Ah, thank you. Makes sense.

**Comment by johncrossman** [comment]
I posted a few questions. Nice work @lyttam

File: src/api/notes.ts (hunk 3 above)
Comment (lyttam) [reply src/api/notes.ts]: Thanks; will amend.

File: tests/test_api/test_notes_controller.py (hunk 2 above)
Comment (lyttam) [reply tests/test_api/test_notes_controller.py]: @johncrossman the autocomplete suggestions are in title case, but the added topics are all caps. I'm going to work on the autocomplete Jira next and I think I'll wind up changing it so topics are stored in title case, but added topics are still displayed in all caps using CSS. It's just a matter of finding the right library to properly convert articles (of, and, or, but) and acronyms (PNP) to title case.

image

File: src/components/note/NewNoteModal.vue (hunk 4 above)
Comment (lyttam) [reply src/components/note/NewNoteModal.vue]: @johncrossman yes, I thought about it. I'll make it shared.

File: tests/test_api/test_notes_controller.py (hunk 2 above)
Comment (johncrossman) [reply tests/test_api/test_notes_controller.py]: Makes sense @lyttam  Ping me when amendments are in and I'll merge.

**Comment by lyttam** [comment]
Ready for re-review, @johncrossman.

**Comment by johncrossman** [comment]
Thanks @lyttam!

**Pull request merged**